File: Public/Proxy/Routers/video.py
# Bu araç @keyiflerolsun tarafından | @KekikAkademi için yazılmıştır.

from CLI                  import konsol
from fastapi              import Request, Response
from starlette.background import BackgroundTask
from fastapi.responses    import StreamingResponse
from .                    import proxy_router
from ..Libs.helpers       import prepare_request_headers, prepare_response_headers, detect_hls_from_url, stream_wrapper, rewrite_hls_manifest
from ..Libs.segment_cache import segment_cache
from urllib.parse         import unquote
import httpx

def is_hls_segment(url: str) -> bool:
    """URL'nin HLS segment'i olup olmadığını kontrol et"""
    url_lower = url.lower()

    # Manifest'leri hariç tut
    if ".m3u8" in url_lower:
        return False

    # Segment göstergeleri
    segment_indicators = (".ts", ".m4s", "seg-", "chunk-", "fragment", ".png")
    return any(indicator in url_lower for indicator in segment_indicators)

@proxy_router.get("/video")
@proxy_router.head("/video")
async def video_proxy(request: Request, url: str, referer: str = None, user_agent: str = None):
    """Video proxy endpoint'i"""
    decoded_url     = unquote(url)
    request_headers = prepare_request_headers(request, decoded_url, referer, user_agent)

    # HLS segment ise cache'i kontrol et
    if is_hls_segment(decoded_url):
        cached_content = await segment_cache.get(decoded_url)
        if cached_content:
            # konsol.print(f"[green]✓ Cache HIT:[/green] {decoded_url[-50:]}")
            return Response(
                content     = cached_content,
                status_code = 200,
                headers     = {
                    "Content-Type"                : "video/MP2T" if decoded_url.endswith('.ts') else "video/iso.segment",
                    "Cache-Control"               : "public, max-age=30",
                    "Access-Control-Allow-Origin" : "*",
                },
            )

    # Client oluştur (SSL doğrulaması devre dışı - bazı sunucular self-signed sertifika kullanıyor)
    client = httpx.AsyncClient(
        follow_redirects = True,
        timeout          = httpx.Timeout(connect=10.0, read=60.0, write=10.0, pool=10.0),
        verify           = False,
    )

    try:
        # HLS Tahmini (URL'den)
        is_hls = detect_hls_from_url(decoded_url)
        detected_content_type = "application/vnd.apple.mpegurl" if is_hls else None

        # GET isteğini başlat
        req = client.build_request("GET", decoded_url, headers=request_headers)
        response = await client.send(req, stream=True)

        if response.status_code >= 400:
            await response.aclose()
            await client.aclose()
            return Response(status_code=response.status_code, content=f"Upstream Error: {response.status_code}")

        # Response headerlarını hazırla
        final_headers = prepare_response_headers(dict(response.headers), decoded_url, detected_content_type)

        # HEAD isteği ise stream yapma, kapat ve dön
        if request.method == "HEAD":
            await response.aclose()
            await client.aclose()
            return Response(
                content     = b"",
                status_code = response.status_code,
                headers     = final_headers,
                media_type  = final_headers.get("Content-Type")
            )

        # HLS manifest ise içeriği yeniden yaz
        if is_hls:
            # Tüm içeriği oku
            content = await response.aread()
            await response.aclose()
            await client.aclose()

            # Manifest URL'lerini yeniden yaz
            rewritten_content = rewrite_hls_manifest(content, decoded_url, referer, user_agent)

            # Content-Length güncelle
            final_headers["Content-Length"] = str(len(rewritten_content))

            return Response(
                content     = rewritten_content,
                status_code = response.status_code,
                headers     = final_headers,
                media_type  = final_headers.get("Content-Type")
            )

        # HLS segment ise cache'e al
        if is_hls_segment(decoded_url):
            content = await response.aread()
            await response.aclose()
            await client.aclose()

            # Cache'e ekle
            await segment_cache.set(decoded_url, content)
            # konsol.print(f"[yellow]⚡ Cache MISS:[/yellow] {decoded_url[-50:]} ({len(content) // 1024}KB)")

            return Response(
                content     = content,
                status_code = response.status_code,
                headers     = final_headers,
                media_type  = final_headers.get("Content-Type")
            )

        # Normal video - StreamingResponse döndür
        return StreamingResponse(
            stream_wrapper(response),
            status_code = response.status_code,
            headers     = final_headers,
            media_type  = final_headers.get("Content-Type"),
            background  = BackgroundTask(client.aclose)
        )

    except Exception as e:
        await client.aclose()
        konsol.print(f"[red]Proxy başlatma hatası: {str(e)}[/red]")
        return Response(status_code=502, content=f"Proxy Error: {str(e)}")

File: services/proxy/handlers/video.go
// Bu araç @keyiflerolsun tarafından | @KekikAkademi için yazılmıştır.

package handlers

import (
	"io"
	"kekik-proxy/cache"
	"kekik-proxy/utils"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/pterm/pterm"
)

var httpClient = &http.Client{
	Timeout: 30 * time.Second,
}

// VideoProxy video proxy endpoint'i
func VideoProxy(c *gin.Context) {
	urlParam := c.Query("url")
	referer := c.Query("referer")
	userAgent := c.Query("user_agent")

	if urlParam == "" {
		c.String(http.StatusBadRequest, "URL parametresi gerekli")
		return
	}

	decodedURL, err := url.QueryUnescape(urlParam)
	if err != nil {
		c.String(http.StatusBadRequest, "Geçersiz URL")
		return
	}

	// Cache kontrolü
	if cachedData, ok := cache.GlobalCache.Get(decodedURL); ok {
		pterm.Debug.Printf("Cache Hit: %s\n", decodedURL)
		c.Data(http.StatusOK, "video/mp2t", cachedData)
		return
	}

	// Request headers hazırla
	reqHeaders := utils.PrepareRequestHeaders(referer, userAgent)

	// Upstream request oluştur
	req2, err := http.NewRequest("GET", decodedURL, nil)
	if err != nil {
		c.String(http.StatusBadGateway, "Request oluşturulamadı: "+err.Error())
		return
	}

	// Range header kopyalama (Safari/iOS desteği için)
	if rangeHeader := c.GetHeader("Range"); rangeHeader != "" {
		req2.Header.Set("Range", rangeHeader)
	}

	// Headers kopyala
	for k, v := range reqHeaders {
		req2.Header[k] = v
	}

	// Upstream'e istek yap
	resp, err := httpClient.Do(req2)
	if err != nil {
		pterm.Error.Printf("Upstream Error (Video): %s -> %v\n", decodedURL, err)
		c.String(http.StatusBadGateway, "Proxy Error: "+err.Error())
		return
	}
	defer resp.Body.Close()

	// Response headers kopyala
	respHeaders := make(http.Header)
	for k, v := range resp.Header {
		if k == "Content-Type" || k == "Content-Length" || k == "Content-Range" || k == "Accept-Ranges" {
			respHeaders[k] = v
		}
	}

	// HLS segmenti ise cache'le
	if strings.HasSuffix(decodedURL, ".ts") || strings.Contains(decodedURL, "seg-") {
		body, err := io.ReadAll(resp.Body)
		if err == nil {
			cache.GlobalCache.Set(decodedURL, body)
			pterm.Debug.Printf("Cache Set: %s\n", decodedURL)
			for k, v := range respHeaders {
				c.Header(k, v[0])
			}
			c.Data(http.StatusOK, "video/mp2t", body)
			return
		}
	}

	// Response
	for k, v := range respHeaders {
		c.Header(k, v[0])
	}
	c.Status(resp.StatusCode)

	// Stream body
	c.Stream(func(w io.Writer) bool {
		buf := make([]byte, 128*1024)
		n, _ := resp.Body.Read(buf)
		if n > 0 {
			w.Write(buf[:n])
			return true
		}
		return false
	})
}
